checks/evaluation/cii_best_practices.go

```diff
@@ -15,48 +15,93 @@
 package evaluation
 
 import (
-	"fmt"
-
 	"github.com/ossf/scorecard/v4/checker"
-	"github.com/ossf/scorecard/v4/clients"
 	sce "github.com/ossf/scorecard/v4/errors"
+	"github.com/ossf/scorecard/v4/finding"
+	"github.com/ossf/scorecard/v4/probes/hasGoldBadge"
+	"github.com/ossf/scorecard/v4/probes/hasInProgressBadge"
+	"github.com/ossf/scorecard/v4/probes/hasPassingBadge"
+	"github.com/ossf/scorecard/v4/probes/hasSilverBadge"
+	"github.com/ossf/scorecard/v4/probes/hasUnknownBadge"
 )
 
-// Note: exported for unit tests.
 const (
 	silverScore = 7
 	// Note: if this value is changed, please update the action's threshold score
 	// https://github.com/ossf/scorecard-action/blob/main/policies/template.yml#L61.
 	passingScore    = 5
 	inProgressScore = 2
+	maxScore        = 10
+	minScore        = 0
+	errScore        = -1
 )
 
 // CIIBestPractices applies the score policy for the CIIBestPractices check.
-func CIIBestPractices(name string, _ checker.DetailLogger, r *checker.CIIBestPracticesData) checker.CheckResult {
-	if r == nil {
-		e := sce.WithMessage(sce.ErrScorecardInternal, "empty raw data")
+func CIIBestPractices(name string,
+	findings []finding.Finding, dl checker.DetailLogger,
+) checker.CheckResult {
+	expectedProbes := []string{
+		hasGoldBadge.Probe,
+		hasSilverBadge.Probe,
+		hasInProgressBadge.Probe,
+		hasPassingBadge.Probe,
+		hasUnknownBadge.Probe,
+	}
+
+	if !finding.UniqueProbesEqual(findings, expectedProbes) {
+		e := sce.WithMessage(sce.ErrScorecardInternal, "invalid probe results")
+		return checker.CreateRuntimeErrorResult(name, e)
+	}
+
+	var countBadges int
+	// Sanity check that we don't have multiple positives
+	for i := range findings {
+		f := &findings[i]
+
+		if f.Outcome == finding.OutcomePositive {
+			countBadges++
+		}
+	}
```

Comment:
We could though. I think there's some discussion in terms of how this should be structured as a probe. Right now there are 5 probes.

1. One alternative could be a single probe who's job is to return the badge level out of the 5 options. This one is hard with the current probe format, and may be hard for policies.

2. The way you have it written now, where there are 5 probes, and only one (the highest?) is set to positive.

3. Keep all 5 probes, but think of it as a minimum. E.g. if you have the silver level, then  inprogress, passing, and silver all return `OutcomePositive`. This makes it easier for people to write policies such as: "i care about repos that are passing or better"

Reply:
Also, do we need a probe for unknown? It's effectively an error state right? Should we delete the dedicated probe and just have the other ones return `OutcomeError`?

Reply:
> Keep all 5 probes, but think of it as a minimum. E.g. if you have the silver level, then inprogress, passing, and silver all return OutcomePositive. This makes it easier for people to write policies such as: "i care about repos that are passing or better"

I like it, but if consumers run the probes in their own policies, they would need to run all probes and consider the highest in their policy. With a probe per badge level, we can only consider one badge level in each probe.

> One alternative could be a single probe who's job is to return the badge level out of the 5 options. This one is hard with the current probe format, and may be hard for policies.

I think I prefer this approach. I imagine the goal with the probes will _always_ be to see whether the project has a badge at a certain level as a minimum. Users will _never_ specifically want to only check whether the project has the InProgress badge. I can't think of reasonable exceptions to that.

Reply:
> > One alternative could be a single probe who's job is to return the badge level out of the 5 options. This one is hard with the current probe format, and may be hard for policies.
> 
> I think I prefer this approach. I imagine the goal with the probes will _always_ be to see whether the project has a badge at a certain level as a minimum. Users will _never_ specifically want to only check whether the project has the InProgress badge. I can't think of reasonable exceptions to that.

I like this approach too.

Reply:
If we have those levels be `int`s we lose some readability, which tier is which number, etc.
Some of that can be addressed in the code having them be consts (e.g. `const Passing = 1`), but in the `probe` format it might be confusing seeing them as ints.

Reply:
This might be an issue in case new badges are introduced that represent levels between the current ones.

Reply:
I've heard before that folks are OK with integers to represent strings, and in fact I was specifically asked to use `int` instead of `strings`, e.g. for https://github.com/ossf/scorecard/blob/main/finding/finding.go#L59 and https://github.com/ossf/scorecard/blob/main/finding/probe/probe.go#L31 (@justaugustus @naveensrinivasan ). I don't remember what the exact arguments were, but I recall something around "easier to match on integers than strings" when writing policies.
 
So what I did is introduce some "space" between integer values, in case we need to add more levels, see https://github.com/ossf/scorecard/blob/main/finding/finding.go#L59. 

Given that probe data is aimed at being interpreted by tools, I think it's OK to have `ints` if we're OK with the "space" solution above.

 Could also ask in the bi-weekly meetings.

Reply:
> I've heard before that folks are OK with integers to represent strings, and in fact I was specifically asked to use int instead of strings

I thought that feedback was in the scenario of a unique probe identifier, in place of a name. Something that isn't ordered, and won't have a collision. Think static analysis tool rule ID: For example from `gosec`:
> G101: Look for hard coded credentials

Reply:
I dont have a strong opinion, but that that mean we need to support for `int` and `string` for the `Value` map.. `Value map[interface{}]`?

Reply:
Not necessarily. (and I think you meant `Value map[string]interface{}` or `Value map[string]any` since Go 1.18)

We could have the level be the key, and just ignore the value, but that seems a little hacky.

E.g. if you have passing, then "passing" would be a key in the value map. Or "gold" if you have gold.
Note: this could factor in with SAST/fuzzing as well, where the only difference is the string

Reply:
> Not necessarily. (and I think you meant `Value map[string]interface{}` or `Value map[string]any` since Go 1.18)
> 
> We could have the level be the key, and just ignore the value, but that seems a little hacky.
> 
> E.g. if you have passing, then "passing" would be a key in the value map. Or "gold" if you have gold. Note: this could factor in with SAST/fuzzing as well, where the only difference is the string

So the probe would have keys in the map corresponding to the badges that the project has?

Reply:
> So the probe would have keys in the `Values` map corresponding to the badges that the project has

or singular value yeah, for the corresponding level, unsure how this would factor in to policy

> and the value are irrelevant?

yes, in this case.

Reply:
Combined the probes into one that returns true if the project has an OpenSSF badge in https://github.com/ossf/scorecard/pull/3520/commits/acbc6d64914699cad04702afc7c078d0f31f33f4. The probe returns the badge level in the `Values`.